File: OpenEdXMobile/src/main/java/org/edx/mobile/view/MyCoursesListFragment.java
package org.edx.mobile.view;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.joanzapata.iconify.fonts.FontAwesomeIcons;

import org.edx.mobile.R;
import org.edx.mobile.core.IEdxEnvironment;
import org.edx.mobile.databinding.FragmentMyCoursesListBinding;
import org.edx.mobile.databinding.PanelFindCourseBinding;
import org.edx.mobile.event.EnrolledInCourseEvent;
import org.edx.mobile.event.MainDashboardRefreshEvent;
import org.edx.mobile.event.NetworkConnectivityChangeEvent;
import org.edx.mobile.exception.AuthException;
import org.edx.mobile.http.HttpStatus;
import org.edx.mobile.http.HttpStatusException;
import org.edx.mobile.http.notifications.FullScreenErrorNotification;
import org.edx.mobile.interfaces.RefreshListener;
import org.edx.mobile.loader.AsyncTaskResult;
import org.edx.mobile.loader.CoursesAsyncLoader;
import org.edx.mobile.logger.Logger;
import org.edx.mobile.model.api.EnrolledCoursesResponse;
import org.edx.mobile.module.db.DataCallback;
import org.edx.mobile.module.prefs.LoginPrefs;
import org.edx.mobile.util.NetworkUtil;
import org.edx.mobile.util.UiUtil;
import org.edx.mobile.view.adapters.MyCoursesAdapter;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import de.greenrobot.event.EventBus;

public class MyCoursesListFragment extends OfflineSupportBaseFragment
        implements RefreshListener,
        LoaderManager.LoaderCallbacks<AsyncTaskResult<List<EnrolledCoursesResponse>>> {

    private static final int MY_COURSE_LOADER_ID = 0x905000;

    private MyCoursesAdapter adapter;
    private FragmentMyCoursesListBinding binding;
    private final Logger logger = new Logger(getClass().getSimpleName());
    private boolean refreshOnResume = false;

    @Inject
    private IEdxEnvironment environment;

    @Inject
    private LoginPrefs loginPrefs;

    private FullScreenErrorNotification errorNotification;

    //TODO: All these callbacks aren't essentially part of MyCoursesListFragment and should move in
    // the Tabs container fragment that's going to be implemented in LEARNER-3251

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        adapter = new MyCoursesAdapter(getActivity(), environment) {
            @Override
            public void onItemClicked(EnrolledCoursesResponse model) {
                environment.getRouter().showCourseDashboardTabs(getActivity(), model, false);
            }

            @Override
            public void onAnnouncementClicked(EnrolledCoursesResponse model) {
                environment.getRouter().showCourseDashboardTabs(getActivity(), model, true);
            }
        };
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_my_courses_list, container, false);
        errorNotification = new FullScreenErrorNotification(binding.myCourseList);
        binding.swipeContainer.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                // Hide the progress bar as swipe layout has its own progress indicator
                binding.loadingIndicator.getRoot().setVisibility(View.GONE);
                errorNotification.hideError();
                loadData(false);
            }
        });
        UiUtil.setSwipeRefreshLayoutColors(binding.swipeContainer);
        // Add empty view to cause divider to render at the top of the list.
        binding.myCourseList.addHeaderView(new View(getContext()), null, false);
        binding.myCourseList.setAdapter(adapter);
        binding.myCourseList.setOnItemClickListener(adapter);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        loadData(true);
    }

    @Override
    public Loader<AsyncTaskResult<List<EnrolledCoursesResponse>>> onCreateLoader(int i, Bundle bundle) {
        return new CoursesAsyncLoader(getActivity());
    }

    @Override
    public void onLoadFinished(Loader<AsyncTaskResult<List<EnrolledCoursesResponse>>> asyncTaskResultLoader, AsyncTaskResult<List<EnrolledCoursesResponse>> result) {
        adapter.clear();
        final Exception exception = result.getEx();
        if (exception != null) {
            if (exception instanceof AuthException) {
                loginPrefs.clear();
                getActivity().finish();
            } else if (exception instanceof HttpStatusException) {
                final HttpStatusException httpStatusException = (HttpStatusException) exception;
                switch (httpStatusException.getStatusCode()) {
                    case HttpStatus.UNAUTHORIZED: {
                        environment.getRouter().forceLogout(getContext(),
                                environment.getAnalyticsRegistry(),
                                environment.getNotificationDelegate());
                        break;
                    }
                }
            } else {
                logger.error(exception);
            }

            errorNotification.showError(getActivity(), exception, R.string.lbl_reload,
                    new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            if (NetworkUtil.isConnected(getContext())) {
                                onRefresh();
                            }
                        }
                    });
        } else if (result.getResult() != null) {
            ArrayList<EnrolledCoursesResponse> newItems = new ArrayList<EnrolledCoursesResponse>(result.getResult());

            updateDatabaseAfterDownload(newItems);

            if (result.getResult().size() > 0) {
                adapter.setItems(newItems);
            }
            addFindCoursesFooter();
            adapter.notifyDataSetChanged();

            if (adapter.isEmpty() && !environment.getConfig().getCourseDiscoveryConfig().isCourseDiscoveryEnabled()) {
                errorNotification.showError(R.string.no_courses_to_display,
                        FontAwesomeIcons.fa_exclamation_circle, 0, null);
                binding.myCourseList.setVisibility(View.GONE);
            } else {
                binding.myCourseList.setVisibility(View.VISIBLE);
                errorNotification.hideError();
            }
        }
        binding.swipeContainer.setRefreshing(false);
        binding.loadingIndicator.getRoot().setVisibility(View.GONE);

        if (!EventBus.getDefault().isRegistered(MyCoursesListFragment.this)) {
            EventBus.getDefault().registerSticky(MyCoursesListFragment.this);
        }
    }

    public void updateDatabaseAfterDownload(ArrayList<EnrolledCoursesResponse> list) {
        if (list != null && list.size() > 0) {
            //update all videos in the DB as Deactivated
            environment.getDatabase().updateAllVideosAsDeactivated(dataCallback);

            for (int i = 0; i < list.size(); i++) {
                //Check if the flag of isIs_active is marked to true,
                //then activate all videos
                if (list.get(i).isIs_active()) {
                    //update all videos for a course fetched in the API as Activated
                    environment.getDatabase().updateVideosActivatedForCourse(list.get(i).getCourse().getId(),
                            dataCallback);
                } else {
                    list.remove(i);
                }
            }

            //Delete all videos which are marked as Deactivated in the database
            environment.getStorage().deleteAllUnenrolledVideos();
        }
    }

    private DataCallback<Integer> dataCallback = new DataCallback<Integer>() {
        @Override
        public void onResult(Integer result) {
        }

        @Override
        public void onFail(Exception ex) {
            logger.error(ex);
        }
    };

    @Override
    public void onLoaderReset(Loader<AsyncTaskResult<List<EnrolledCoursesResponse>>> asyncTaskResultLoader) {
        adapter.clear();
        adapter.notifyDataSetChanged();
        binding.myCourseList.setVisibility(View.GONE);
        binding.loadingIndicator.getRoot().setVisibility(View.VISIBLE);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (refreshOnResume) {
            loadData(true);
            refreshOnResume = false;
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }

    @SuppressWarnings("unused")
    public void onEventMainThread(EnrolledInCourseEvent event) {
        refreshOnResume = true;
    }

    protected void loadData(boolean showProgress) {
        if (showProgress) {
            binding.loadingIndicator.getRoot().setVisibility(View.VISIBLE);
            errorNotification.hideError();
        }
        getLoaderManager().restartLoader(MY_COURSE_LOADER_ID, null, this);
    }

    private void addFindCoursesFooter() {
        // Validate footer is not already added.
        if (binding.myCourseList.getFooterViewsCount() > 0) {
            return;
        }
        if (environment.getConfig().getCourseDiscoveryConfig().isCourseDiscoveryEnabled()) {
            // Add 'Find a Course' list item as a footer.
            final PanelFindCourseBinding footer = DataBindingUtil.inflate(LayoutInflater.from(getActivity()),
                    R.layout.panel_find_course, binding.myCourseList, false);
            binding.myCourseList.addFooterView(footer.getRoot(), null, false);
            footer.courseBtn.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    environment.getAnalyticsRegistry().trackUserFindsCourses();
                    environment.getRouter().showFindCourses(getActivity());
                }
            });
        }
        // Add empty view to cause divider to render at the bottom of the list.
        binding.myCourseList.addFooterView(new View(getContext()), null, false);
    }

    @Override
    public void onRefresh() {
        EventBus.getDefault().post(new MainDashboardRefreshEvent());
    }

    @SuppressWarnings("unused")
    public void onEvent(MainDashboardRefreshEvent event) {
        loadData(true);
    }

    @Override
    protected void onRevisit() {
        super.onRevisit();
        if (NetworkUtil.isConnected(getActivity())) {
            binding.swipeContainer.setEnabled(true);
        }
    }

    @SuppressWarnings("unused")
    public void onEvent(NetworkConnectivityChangeEvent event) {
        if (getActivity() != null) {
            if (NetworkUtil.isConnected(getContext())) {
                binding.swipeContainer.setEnabled(true);
            } else {
                //Disable swipe functionality and hide the loading view
                binding.swipeContainer.setEnabled(false);
                binding.swipeContainer.setRefreshing(false);
            }
            onNetworkConnectivityChangeEvent(event);
        }
    }

    @Override
    protected boolean isShowingFullScreenError() {
        return errorNotification != null && errorNotification.isShowing();
    }
}

File: OpenEdXMobile/src/main/java/org/edx/mobile/view/MyCoursesListFragment.kt
package org.edx.mobile.view

import android.content.DialogInterface
import android.os.Bundle
import android.os.SystemClock
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver.OnScrollChangedListener
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import dagger.hilt.android.AndroidEntryPoint
import org.edx.mobile.R
import org.edx.mobile.authentication.LoginAPI
import org.edx.mobile.base.MainApplication
import org.edx.mobile.databinding.FragmentMyCoursesListBinding
import org.edx.mobile.deeplink.DeepLink
import org.edx.mobile.deeplink.DeepLinkManager
import org.edx.mobile.deeplink.Screen
import org.edx.mobile.event.EnrolledInCourseEvent
import org.edx.mobile.event.IAPFlowEvent
import org.edx.mobile.event.MainDashboardRefreshEvent
import org.edx.mobile.event.MoveToDiscoveryTabEvent
import org.edx.mobile.event.MyCoursesRefreshEvent
import org.edx.mobile.event.NetworkConnectivityChangeEvent
import org.edx.mobile.exception.ErrorMessage
import org.edx.mobile.extenstion.parcelable
import org.edx.mobile.extenstion.setVisibility
import org.edx.mobile.http.HttpStatus
import org.edx.mobile.http.HttpStatusException
import org.edx.mobile.http.notifications.FullScreenErrorNotification
import org.edx.mobile.http.notifications.SnackbarErrorNotification
import org.edx.mobile.interfaces.RefreshListener
import org.edx.mobile.model.api.EnrolledCoursesResponse
import org.edx.mobile.model.iap.IAPFlowData
import org.edx.mobile.module.analytics.Analytics
import org.edx.mobile.module.analytics.InAppPurchasesAnalytics
import org.edx.mobile.util.AppStoreUtils
import org.edx.mobile.util.InAppPurchasesException
import org.edx.mobile.util.NetworkUtil
import org.edx.mobile.util.NonNullObserver
import org.edx.mobile.util.UiUtils
import org.edx.mobile.util.observer.EventObserver
import org.edx.mobile.view.adapters.BaseListAdapter
import org.edx.mobile.view.adapters.MyCoursesListAdapter
import org.edx.mobile.view.custom.error.EdxErrorState
import org.edx.mobile.view.dialog.CourseModalDialogFragment
import org.edx.mobile.view.dialog.FullscreenLoaderDialogFragment
import org.edx.mobile.viewModel.CourseViewModel
import org.edx.mobile.viewModel.CourseViewModel.CoursesRequestType
import org.edx.mobile.viewModel.InAppPurchasesViewModel
import org.edx.mobile.wrapper.InAppPurchasesDialog
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import javax.inject.Inject

@AndroidEntryPoint
class MyCoursesListFragment : OfflineSupportBaseFragment(), RefreshListener {

    private lateinit var adapter: MyCoursesListAdapter
    private lateinit var binding: FragmentMyCoursesListBinding

    private val courseViewModel: CourseViewModel by viewModels()
    private val iapViewModel: InAppPurchasesViewModel by viewModels()

    @Inject
    lateinit var loginAPI: LoginAPI

    @Inject
    lateinit var iapAnalytics: InAppPurchasesAnalytics

    @Inject
    lateinit var iapDialog: InAppPurchasesDialog

    private lateinit var errorNotification: FullScreenErrorNotification
    private lateinit var onScrollChangedListener: OnScrollChangedListener

    private var refreshOnResume = false
    private var isObserversInitialized = true
    private var lastClickTime: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        adapter = object : MyCoursesListAdapter(environment) {
            override fun onItemClicked(model: EnrolledCoursesResponse) {
                activity?.let { activity ->
                    environment.router.showCourseDashboardTabs(activity, model)
                }
            }

            override fun onAnnouncementClicked(model: EnrolledCoursesResponse) {
                activity?.let { activity ->
                    environment.router.showCourseDashboardTabs(
                        activity,
                        model,
                        Screen.COURSE_ANNOUNCEMENT
                    )
                }
            }

            override fun onValuePropClicked(model: EnrolledCoursesResponse) {
                //This time is checked to avoid taps in quick succession
                val currentTime = SystemClock.elapsedRealtime()
                if (currentTime - lastClickTime > BaseListAdapter.MIN_CLICK_INTERVAL) {
                    lastClickTime = currentTime
                    CourseModalDialogFragment.newInstance(
                        Analytics.Screens.COURSE_ENROLLMENT,
                        model.courseId,
                        model.courseSku,
                        model.course.name,
                        model.course.isSelfPaced
                    ).show(childFragmentManager, CourseModalDialogFragment.TAG)
                }
            }
        }
        detectDeeplink()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentMyCoursesListBinding.inflate(inflater, container, false)
        errorNotification = FullScreenErrorNotification(binding.myCourseList)

        binding.swipeContainer.setOnRefreshListener {
            errorNotification.hideError()
            binding.stateLayout.root.setVisibility(false)
            courseViewModel.fetchEnrolledCourses(
                type = CoursesRequestType.STALE,
                showProgress = false
            )
        }
        UiUtils.setSwipeRefreshLayoutColors(binding.swipeContainer)

        binding.myCourseList.adapter = adapter
        binding.myCourseList.addItemDecoration(
            DividerItemDecoration(requireContext(), LinearLayoutManager.VERTICAL).apply {
                setDrawable(
                    UiUtils.getDrawable(
                        requireContext(),
                        R.drawable.my_course_list_recycler_view_divider
                    )
                )
            }
        )
        initCourseObservers()
        courseViewModel.fetchEnrolledCourses(type = CoursesRequestType.PERSISTABLE_CACHE)

        return binding.root
    }

    private fun initCourseObservers() {
        courseViewModel.showProgress.observe(viewLifecycleOwner, NonNullObserver {
            binding.loadingIndicator.root.setVisibility(it)
            if (it) {
                errorNotification.hideError()
                binding.stateLayout.root.setVisibility(false)
            }
        })

        courseViewModel.enrolledCoursesResponse.observe(
            viewLifecycleOwner,
            EventObserver { enrolledCourses ->
                populateCourseData(data = enrolledCourses)
                // Checking if the user is logged-in as we need userId to enable IAP
                if (environment.loginPrefs.isUserLoggedIn &&
                    environment.featuresPrefs.isIAPEnabledForUser(environment.loginPrefs.isOddUserId)
                ) {
                    initInAppPurchaseSetup()
                    val fullscreenLoader = FullscreenLoaderDialogFragment.getRetainedInstance(
                        fragmentManager = childFragmentManager
                    )
                    if (fullscreenLoader != null) {
                        SnackbarErrorNotification(binding.root).showUpgradeSuccessSnackbar(R.string.purchase_success_message)
                        fullscreenLoader.closeLoader()
                    } else if (environment.featuresPrefs.canAutoCheckUnfulfilledPurchase &&
                        courseViewModel.courseRequestType != CoursesRequestType.PERSISTABLE_CACHE
                    ) {
                        detectUnfulfilledPurchase(enrolledCourses)
                    }
                }
            })

        courseViewModel.handleError.observe(viewLifecycleOwner, NonNullObserver {
            when {
                (it is HttpStatusException) -> {
                    when (it.statusCode) {
                        HttpStatus.UNAUTHORIZED -> {
                            context?.let { context ->
                                environment.router?.forceLogout(
                                    context,
                                    environment.analyticsRegistry,
                                    environment.notificationDelegate
                                )
                            }
                        }

                        HttpStatus.UPGRADE_REQUIRED -> {
                            showError(EdxErrorState.State.UPDATE_APP)
                        }

                        else -> {
                            showError(EdxErrorState.State.NETWORK)
                        }
                    }
                }

                (FullscreenLoaderDialogFragment
                    .getRetainedInstance(fragmentManager = childFragmentManager)?.isAdded == true) -> {
                    iapViewModel.dispatchError(
                        requestType = ErrorMessage.COURSE_REFRESH_CODE,
                        errorMessage = it.message
                    )
                }

                else -> {
                    showError(EdxErrorState.State.NETWORK)
                }
            }
            invalidateView()
        })
    }

    private fun initIAPObservers() {
        iapViewModel.refreshCourseData.observe(viewLifecycleOwner, EventObserver { iapFlowData ->
            iapDialog.showNewExperienceAlertDialog(this, { _, _ ->
                iapAnalytics.trackIAPEvent(
                    eventName = Analytics.Events.IAP_NEW_EXPERIENCE_ALERT_ACTION,
                    actionTaken = Analytics.Values.ACTION_REFRESH
                )
                iapAnalytics.initUnlockContentTime()
                showFullscreenLoader(iapFlowData)
            }, { _, _ ->
                iapAnalytics.trackIAPEvent(
                    eventName = Analytics.Events.IAP_NEW_EXPERIENCE_ALERT_ACTION,
                    actionTaken = Analytics.Values.ACTION_CONTINUE_WITHOUT_UPDATE
                )
            })
        })

        iapViewModel.errorMessage.observe(viewLifecycleOwner, EventObserver { errorMessage ->
            var retryListener: DialogInterface.OnClickListener? = null
            if (errorMessage.canRetry()) {
                retryListener = DialogInterface.OnClickListener { _, _ ->
                    if (errorMessage.requestType == ErrorMessage.EXECUTE_ORDER_CODE) {
                        iapViewModel.executeOrder()
                    } else if (errorMessage.requestType == ErrorMessage.COURSE_REFRESH_CODE) {
                        courseViewModel.fetchEnrolledCourses(
                            type = CoursesRequestType.LIVE,
                            showProgress = false
                        )
                    } else if (HttpStatus.NOT_ACCEPTABLE == (errorMessage.throwable as InAppPurchasesException).httpErrorCode) {
                        showFullscreenLoader(iapFlowData = iapViewModel.iapFlowData)
                    }
                }
            }
            iapDialog.handleIAPException(
                fragment = this@MyCoursesListFragment,
                errorMessage = errorMessage,
                retryListener = retryListener,
                cancelListener = { _, _ ->
                    iapViewModel.iapFlowData.clear()
                    if (errorMessage.isPostUpgradeErrorType()) {
                        FullscreenLoaderDialogFragment.getRetainedInstance(childFragmentManager)
                            ?.dismiss()
                    }
                }
            )
        })
    }

    /**
     * Method to detect Unfulfilled Purchases.
     *
     * @param enrolledCourses User enrolled courses.
     * */
    private fun detectUnfulfilledPurchase(enrolledCourses: List<EnrolledCoursesResponse>) {
        iapAnalytics.reset()
        iapViewModel.detectUnfulfilledPurchase(
            environment.loginPrefs.userId,
            enrolledCourses,
            IAPFlowData.IAPFlowType.SILENT,
            Analytics.Screens.COURSE_ENROLLMENT
        )
    }

    override fun onStart() {
        super.onStart()
        // To force the SwipeRefreshLayout to use the scroll only when the underlying view has
        // scrolled up to its top.
        binding.swipeContainer.viewTreeObserver.addOnScrollChangedListener(
            OnScrollChangedListener {
                binding.swipeContainer.isEnabled = binding.stateLayout.root.scrollY == 0
            }.also {
                onScrollChangedListener = it
            })
    }

    override fun onResume() {
        super.onResume()
        if (!EventBus.getDefault().isRegistered(this@MyCoursesListFragment)) {
            EventBus.getDefault().register(this@MyCoursesListFragment)
        }
        if (refreshOnResume) {
            courseViewModel.fetchEnrolledCourses(
                type = CoursesRequestType.PERSISTABLE_CACHE,
                showProgress = false
            )
            refreshOnResume = false
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        binding.swipeContainer.viewTreeObserver.removeOnScrollChangedListener(
            onScrollChangedListener
        )
        EventBus.getDefault().unregister(this)
    }

    @Subscribe(sticky = true)
    @Suppress("UNUSED_PARAMETER")
    fun onEventMainThread(event: EnrolledInCourseEvent?) {
        refreshOnResume = true
    }

    @Subscribe
    fun onEventMainThread(event: IAPFlowEvent) {
        if (!this.isResumed) {
            return
        }
        when (event.flowAction) {
            IAPFlowData.IAPAction.SHOW_FULL_SCREEN_LOADER -> {
                event.iapFlowData?.let {
                    showFullscreenLoader(event.iapFlowData)
                }
            }

            IAPFlowData.IAPAction.PURCHASE_FLOW_COMPLETE -> {
                courseViewModel.fetchEnrolledCourses(type = CoursesRequestType.LIVE)
            }
        }
    }

    private fun initInAppPurchaseSetup() {
        if (isAdded && environment.featuresPrefs.isValuePropEnabled && isObserversInitialized) {
            initIAPObservers()
            isObserversInitialized = false
        }
    }

    private fun populateCourseData(
        data: List<EnrolledCoursesResponse>
    ) {
        adapter.submitList(data)

        if (adapter.itemCount == 0 && environment.config.discoveryConfig.isDiscoveryEnabled) {
            binding.stateLayout.state.setState(EdxErrorState.State.EMPTY, Screen.MY_COURSES)
            binding.stateLayout.state.setActionListener {
                environment.analyticsRegistry?.trackUserFindsCourses(adapter.itemCount)
                EventBus.getDefault().post(MoveToDiscoveryTabEvent(Screen.DISCOVERY))
            }
            binding.stateLayout.root.setVisibility(true)
            binding.myCourseList.setVisibility(false)
        } else if (adapter.itemCount == 0 && !environment.config.discoveryConfig.isDiscoveryEnabled) {
            errorNotification.showError(
                R.string.no_courses_to_display,
                R.drawable.ic_error, 0, null
            )
            binding.myCourseList.setVisibility(false)
        } else {
            binding.stateLayout.root.setVisibility(false)
            binding.myCourseList.setVisibility(true)
        }
        invalidateView()
    }

    private fun showFullscreenLoader(iapFlowData: IAPFlowData) {
        // To proceed with the same instance of dialog fragment in case of orientation change
        val fullscreenLoader = try {
            FullscreenLoaderDialogFragment
                .getRetainedInstance(fragmentManager = childFragmentManager) as FullscreenLoaderDialogFragment
        } catch (e: Exception) {
            FullscreenLoaderDialogFragment.newInstance(iapFlowData)
        }
        fullscreenLoader.show(childFragmentManager, FullscreenLoaderDialogFragment.TAG)
    }

    private fun detectDeeplink() {
        arguments?.parcelable<DeepLink>(Router.EXTRA_DEEP_LINK)?.let {
            DeepLinkManager.proceedDeeplink(requireActivity(), it)
            MainApplication.instance().showBanner(loginAPI, true)
            arguments?.putParcelable(Router.EXTRA_DEEP_LINK, null)
        } ?: run {
            MainApplication.instance().showBanner(loginAPI, false)
        }
    }

    private fun invalidateView() {
        binding.swipeContainer.isRefreshing = false
        binding.loadingIndicator.root.visibility = View.GONE
    }

    private fun showError(state: EdxErrorState.State) {
        binding.myCourseList.setVisibility(false)
        binding.stateLayout.root.setVisibility(true)
        binding.stateLayout.state.setState(state, Screen.MY_COURSES)
        binding.stateLayout.state.setActionListener {
            when (state) {
                EdxErrorState.State.UPDATE_APP -> {
                    AppStoreUtils.openAppInAppStore(requireContext())
                }

                EdxErrorState.State.NETWORK -> {
                    onRefresh()
                }

                else -> {}
            }
        }
    }

    override fun onRefresh() {
        EventBus.getDefault().post(MainDashboardRefreshEvent())
    }

    @Subscribe(sticky = true)
    @Suppress("UNUSED_PARAMETER")
    fun onEvent(event: MainDashboardRefreshEvent) {
        courseViewModel.fetchEnrolledCourses(type = CoursesRequestType.LIVE)
    }

    @Subscribe(sticky = true)
    @Suppress("UNUSED_PARAMETER")
    fun onEvent(event: MyCoursesRefreshEvent) {
        courseViewModel.fetchEnrolledCourses(type = CoursesRequestType.LIVE)
    }

    override fun onRevisit() {
        super.onRevisit()
        if (NetworkUtil.isConnected(activity)) {
            binding.swipeContainer.isEnabled = true
        }
    }

    @Subscribe(sticky = true)
    fun onEvent(event: NetworkConnectivityChangeEvent?) {
        if (activity != null) {
            if (NetworkUtil.isConnected(context)) {
                binding.swipeContainer.isEnabled = true
            } else {
                //Disable swipe functionality and hide the loading view
                binding.swipeContainer.isEnabled = false
                binding.swipeContainer.isRefreshing = false
            }
            onNetworkConnectivityChangeEvent(event)
        }
    }

    override fun isShowingFullScreenError(): Boolean {
        return errorNotification.isShowing
    }
}
